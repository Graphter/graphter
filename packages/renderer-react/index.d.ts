import ListRenderer from './list-renderer';
export { ListRenderer };
