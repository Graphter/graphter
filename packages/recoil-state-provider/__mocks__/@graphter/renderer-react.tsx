import { NodeRendererRegistration } from "@graphter/core";
import React from "react";

export const nodeRendererStore = {
  get: jest.fn()
}

export const pathConfigStore = {
  get: jest.fn(),
  set: jest.fn()
}

export const NodeValidationProvider = jest.fn(({ children }: any) => <div>{children}</div>)
export const NodeDataProvider = jest.fn(({ children }: any) => <div>{children}</div>)