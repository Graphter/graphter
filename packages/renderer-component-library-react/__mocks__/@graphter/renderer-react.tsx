import React, { useState } from "react";

export const useArrayNodeData = jest.fn(() => {
  return {
    childIds: [],
    removeItem: jest.fn(),
    commitItem: jest.fn()
  }
})

export const useNodeData = jest.fn(() => {
  return [
    null,
    () => {}
  ]
})

export const useNodeValidation = jest.fn().mockResolvedValue([])

export const nodeRendererStore = {
  get: jest.fn((nodeType: string) => ({
    type: 'string',
    renderer: ({originalNodeData}: any) => {
      const [ value, setValue ] = useState(originalNodeData || '')
      return (
        <input value={value} onChange={e => setValue(e.target.value)} data-testid='string-renderer' />
    )
    }
  }))
}
export const createDefault = jest.fn()