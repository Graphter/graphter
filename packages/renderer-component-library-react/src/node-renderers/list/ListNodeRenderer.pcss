@import '@graphter/theme/src/colours.pcss';

.listNodeRenderer{

}
.defaultItemWrapper{
    background-color: rgb(248, 248, 248);
    border-radius: 0.2em;
    padding: 10px;
    margin-bottom: 10px
}
.defaultNewItemWrapper{
    border: 1px solid rgb(234, 234, 234);
    border-radius: 0.2em;
    padding: 10px;
}
.items{

}
.button{
    display:inline-block;
    padding: 1em;
    border-radius: 0.5em;
    font-weight:400;
    color:var(--dark-text);
    background-color:#efefef;
    text-align:center;
    position:relative;
    border: 0 none transparent;
}