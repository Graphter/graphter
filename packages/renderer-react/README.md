# Graphter React renderer

A React framework that, when combined with node renderers such as those provided in the [Component Lib](../renderer-component-library-react/README.md) package, allows humans to manage arbitrary data structures.