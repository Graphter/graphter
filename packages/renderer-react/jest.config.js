module.exports = {
  preset: 'ts-jest',
  testEnvironment: 'jsdom',
  moduleNameMapper: {
    "\\.(jpg|jpeg|png|gif|eot|otf|webp|svg|ttf|woff|woff2|mp4|webm|wav|mp3|m4a|aac|oga)$": "<rootDir>/__mocks__/fileMock.ts",
    "\\.(pcss|css|less)$": "identity-obj-proxy"
  },
  setupFilesAfterEnv: [ './setupTests.ts' ],
  automock: true,
  unmockedModulePathPatterns: [],
  snapshotSerializers: [
    'enzyme-to-json'
  ],
  testMatch: ['<rootDir>/**/*.test.ts*'],
  testPathIgnorePatterns: ['node_modules'],
};