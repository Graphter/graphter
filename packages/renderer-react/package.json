{
  "name": "malle-renderer-react",
  "version": "1.0.0",
  "description": "",
  "main": "dist/index.js",
  "module": "dist/index.es.js",
  "files": [
    "dist"
  ],
  "types": "dist/index.d.ts",
  "scripts": {
    "test": "cross-env CI=1 react-scripts test --env=jsdom",
    "test:watch": "react-scripts test --env=jsdom",
    "build": "rollup -c",
    "start": "rollup -c -w",
    "prepare": "npm run build"
  },
  "author": "",
  "license": "ISC",
  "peerDependencies": {
    "react": "^16.13.1",
    "react-dom": "^16.13.1"
  },
  "dependencies": {},
  "devDependencies": {
    "react": "^16.13.1",
    "react-dom": "^16.13.1",
    "@babel/core": "^7.9.0",
    "@babel/plugin-proposal-export-default-from": "^7.8.3",
    "@babel/preset-env": "^7.9.0",
    "@babel/preset-react": "^7.9.1",
    "@babel/preset-stage-0": "^7.8.3",
    "@svgr/rollup": "^5.3.0",
    "@types/react": "^16.9.25",
    "eslint": "^6.8.0",
    "react-scripts": "^3.4.1",
    "rollup": "^2.2.0",
    "rollup-plugin-babel": "^4.4.0",
    "rollup-plugin-commonjs": "^10.1.0",
    "rollup-plugin-node-resolve": "^5.2.0",
    "rollup-plugin-peer-deps-external": "^2.2.2",
    "rollup-plugin-postcss": "^2.5.0",
    "rollup-plugin-typescript2": "^0.26.0",
    "rollup-plugin-url": "^3.0.1",
    "typescript": "^3.8.3"
  }
}
