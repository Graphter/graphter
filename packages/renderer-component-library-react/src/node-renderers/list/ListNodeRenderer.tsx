import React, { ComponentType, useState } from 'react'
import { NodeRendererProps } from "@graphter/core";
import { nodeRendererStore, createDefault } from "@graphter/renderer-react";
import { useArrayNodeData } from "@graphter/renderer-react";
import { setupNodeRenderer } from "@graphter/renderer-react";
import DefaultItemView from "./DefaultItemView";
import DefaultNewItemWrapper from "./DefaultNewItemWrapper";
import DefaultEditItemWrapper from "./DefaultEditItemWrapper";
import { pathUtils } from "@graphter/renderer-react";
import { useTreeDataInitialiser } from "@graphter/renderer-react";

const ListNodeRenderer: ComponentType<NodeRendererProps> = setupNodeRenderer((
  {
    config,
    originalTreeData,
    committed,
    globalPath,
    ErrorDisplayComponent,
    options
  }: NodeRendererProps
) => {
  if (!config) throw new Error(`<ListNodeRenderer /> component at '${globalPath.join('/')}' is missing config`)
  const originalNodeData = pathUtils.getValue(originalTreeData, globalPath.slice(2), createDefault(config, []))
  if (!Array.isArray(originalNodeData)) throw new Error(`'${config.type}' renderer only works with arrays but got '${typeof originalNodeData}'`)
  if (!config.children || !config.children.length) throw new Error(`'${config.type}' renderer must have at least one child config`)
  if (config.children.length > 1) throw new Error('Only one child list type is currently supported')

  const [ showNewItemUI, setShowNewItemUI ] = useState(false)
  const [ editingItems, setEditingItems ] = useState<Set<string>>(new Set())

  const childConfig = config.children[0]
  const childRendererRegistration = nodeRendererStore.get(childConfig.type)
  if (!childRendererRegistration) throw new Error(`Couldn't find a renderer for child renderer type ${childConfig.type} at ${globalPath.join('/')}/${childConfig.id}`)
  const ChildTypeRenderer = childRendererRegistration.Renderer

  const treeDataInitialiser = useTreeDataInitialiser()

  const {
    childIds,
    removeItem,
    commitItem,
  } = useArrayNodeData(globalPath)

  return (
    <div className='flex flex-col' data-nodetype='list' data-nodepath={globalPath.join('/')}>
      <div className='' data-testid='items'>
        {childIds && childIds.map((childId: any, i: number) => {
          const childPath = [ ...globalPath, i ]
          if (!editingItems.has(childId)) return (
            <DefaultItemView
              key={childId}
              childId={childId}
              config={childConfig}
              globalPath={childPath}
              onSelect={() => {
                if (config.options?.itemSelectionBehaviour === 'INLINE' || !config.options?.itemSelectionBehaviour) {
                  const newEditingItems = new Set(editingItems)
                  newEditingItems.add(childId)
                  setEditingItems(newEditingItems)
                  return
                }
                if (config.options?.itemSelectionBehaviour === 'CUSTOM' && typeof options?.customItemSelectionBehaviour === 'function') {
                  options.customItemSelectionBehaviour(config.options?.itemSelectionBehaviour, childConfig, childPath)
                }
              }}
              onRemove={() => {
                removeItem(i)
              }}
            />
          )
          return (
            <DefaultEditItemWrapper
              key={childId}
              onRemove={() => {
                removeItem(i)
              }}
              onDone={() => {
                const newEditingItems = new Set(editingItems)
                newEditingItems.delete(childId)
                setEditingItems(newEditingItems)
              }}
            >
              <ChildTypeRenderer
                config={childConfig}
                globalPath={childPath}
                committed={committed}
                originalTreeData={originalTreeData}
                ErrorDisplayComponent={ErrorDisplayComponent}
                options={childRendererRegistration.options}
              />
            </DefaultEditItemWrapper>
          )
        })}
      </div>
      {showNewItemUI ? (
        <DefaultNewItemWrapper
          config={config}
          onAdd={() => {
            setShowNewItemUI(false)
            commitItem(childIds.length)
          }}
          onCancel={() => {
            setShowNewItemUI(false)
          }}
        >
          <ChildTypeRenderer
            config={childConfig}
            globalPath={[ ...globalPath, childIds.length ]}
            committed={false}
            originalTreeData={originalTreeData}
            ErrorDisplayComponent={ErrorDisplayComponent}
            options={childRendererRegistration.options}
          />
        </DefaultNewItemWrapper>
      ) : (
        <button
          type='button'
          className='p-5 border border-dashed rounded hover:border-blue-200 hover:bg-gray-50 transition-colours duration-200 text-blue-300'
          onClick={() => {
            treeDataInitialiser(childConfig, [ ...globalPath, childIds.length ], false)
            setShowNewItemUI(true)
          }}
          data-testid='add-item-btn'
        >[+]</button>
      )}

    </div>
  )
})

export default ListNodeRenderer