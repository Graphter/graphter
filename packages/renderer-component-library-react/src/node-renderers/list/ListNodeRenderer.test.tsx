import React, { useState } from "react";
import { act, render, fireEvent, within, queryAllByTestId } from '@testing-library/react';
import * as graphterRenderer from "@graphter/renderer-react";
import ListNodeRenderer from "./ListNodeRenderer";
import clone from 'rfdc'

const useArrayNodeDataMock = graphterRenderer.useArrayNodeData as jest.Mock<any>
const nodeRendererStoreGetMock = graphterRenderer.nodeRendererStore.get as jest.Mock<any>
const createDefaultMock = graphterRenderer.createDefault as jest.Mock<any>

describe('<ListNodeRenderer />', () => {
  beforeEach(() => {
  })
  afterEach(() => {
    jest.clearAllMocks()
  })
  it('should render correctly', () => {
    useArrayNodeDataMock.mockReturnValue({ childIds: [ 'one', 'two' ] })
    const {container} = render(<ListNodeRenderer
      config={{
        id: 'name',
        name: 'Name',
        description: 'The name',
        type: 'list',
        children: [ {
          id: 'list-item',
          name: 'List item',
          type: 'string'
        } ]
      }}
      originalNodeData={[
        'item 1 value',
        'item 2 value'
      ]}
      committed={true}
      path={[ '/' ]}
    />);
    expect(container).toMatchSnapshot();
  })
  it('should render default data when new', () => {
    createDefaultMock.mockReturnValue([])
    const config = {
      id: 'name',
      name: 'Name',
      description: 'The name',
      type: 'list',
      children: [ {
        id: 'list-item',
        name: 'List item',
        type: 'string'
      } ]
    }
    render(<ListNodeRenderer
      config={clone()(config)}
      originalNodeData={undefined}
      committed={true}
      path={[ '/' ]}
    />);
    expect(graphterRenderer.createDefault).toHaveBeenCalledWith(config, [])
    expect(useArrayNodeDataMock).toHaveBeenCalledWith([ '/' ], config, [], true)
  })
  it('should use the array data provider for data', () => {
    useArrayNodeDataMock.mockReturnValue({ childIds: [ 'one', 'two' ] })
    const config = {
      id: 'name',
      name: 'Name',
      description: 'The name',
      type: 'list',
      children: [ {
        id: 'list-item',
        name: 'List item',
        type: 'string'
      } ]
    }
    render(<ListNodeRenderer
      config={clone()(config)}
      originalNodeData={[
        'item 1 value',
        'item 2 value'
      ]}
      committed={true}
      path={[ '/' ]}
    />);
    expect(useArrayNodeDataMock).toHaveBeenCalled()
  })
  it('should throw if non-array data is passed in', () => {
    expect.assertions(1)
    try {
      render(<ListNodeRenderer
        config={{
          id: 'name',
          name: 'Name',
          description: 'The name',
          type: 'list',
          children: [ {
            id: 'list-item',
            name: 'List item',
            type: 'string'
          } ]
        }}
        originalNodeData={{
          some: 'non-array-data'
        }}
        committed={true}
        path={[ '/' ]}
      />);
    } catch (err) {
      expect(err).toMatchSnapshot()
    }
  })
  it('should error when no child config is supplied', () => {
    expect.assertions(1)
    try {
      // @ts-ignore
      render(<ListNodeRenderer config={undefined}
                               originalNodeData={[
                                 'item 1 value',
                                 'item 2 value'
                               ]}
                               committed={true}
                               path={[ '/' ]}
      />);
    } catch (err) {
      expect(err).toMatchSnapshot()
    }
  })
  it('should error when more than one child config is supplied', () => {
    expect.assertions(1)
    try {
      render(<ListNodeRenderer
        config={{
          id: 'name',
          name: 'Name',
          description: 'The name',
          type: 'list',
          children: [
            {
              id: 'list-item-1',
              name: 'List item 1',
              type: 'string'
            },
            {
              id: 'list-item-2',
              name: 'List item 2',
              type: 'string'
            },
          ]
        }}
        originalNodeData={[
          'item 1 value',
          'item 2 value'
        ]}
        committed={true}
        path={[ '/' ]}
      />)
    } catch (err) {
      expect(err).toMatchSnapshot()
    }
  })
  it('should use the correct child renderer', () => {
    render(<ListNodeRenderer
      config={{
        id: 'name',
        name: 'Name',
        description: 'The name',
        type: 'list',
        children: [ {
          id: 'list-item',
          name: 'List item',
          type: 'string'
        } ]
      }}
      originalNodeData={[
        'item 1 value',
        'item 2 value'
      ]}
      committed={true}
      path={[ '/' ]}
    />);
    expect(nodeRendererStoreGetMock).toHaveBeenCalledWith('string')
  })
  it('should render an empty item to add', () => {
    const { queryByTestId, getByText } = render(<ListNodeRenderer
      config={{
        id: 'name',
        name: 'Name',
        description: 'The name',
        type: 'list',
        children: [ {
          id: 'list-item',
          name: 'List item',
          type: 'string'
        } ]
      }}
      originalNodeData={[
        'item 1 value',
        'item 2 value'
      ]}
      committed={true}
      path={[ '/' ]}
    />);
    act(() => {
      fireEvent.click(getByText('[+]'))
    })
    expect(queryByTestId('add-item')).not.toBeNull()
  })
  it('should commit a new item when the "add" button is selected', () => {
    const commitItemMock = jest.fn()
    useArrayNodeDataMock.mockImplementation(() => ({
      childIds: [ '1', '2' ],
      removeItem: () => {},
      commitItem: commitItemMock
    }))
    const { getByTestId, getByText } = render(<ListNodeRenderer
      config={{
        id: 'name',
        name: 'Name',
        description: 'The name',
        type: 'list',
        children: [ {
          id: 'list-item',
          name: 'List item',
          type: 'string'
        } ]
      }}
      originalNodeData={[
        'item 1 value',
        'item 2 value'
      ]}
      committed={true}
      path={[ '/' ]}
    />);
    act(() => {
      fireEvent.click(getByText('[+]'))
    })
    const addItemContainerQueries = within(getByTestId('add-item'))
    act(() => {
      fireEvent.change(addItemContainerQueries.getByDisplayValue(''), {
        target:
          {
            value: 'some new value'
          }
      })
    })
    act(() => {
      fireEvent.click(getByTestId('add-item-btn'))
    })
    expect(commitItemMock).toHaveBeenCalledWith(2)
  })
  it.each([0, 1])
  ('should remove an item at the correct location (%s) when the remove ui is clicked', (index) => {
    const removeItemMock = jest.fn()
    useArrayNodeDataMock.mockImplementation(() => ({
      childIds: [ '1', '2' ],
      removeItem: removeItemMock,
      commitItem: () => {}
    }))
    const { queryAllByTestId } = render(<ListNodeRenderer
      config={{
        id: 'name',
        name: 'Name',
        description: 'The name',
        type: 'list',
        children: [ {
          id: 'list-item',
          name: 'List item',
          type: 'string'
        } ]
      }}
      originalNodeData={[
        'item 1 value',
        'item 2 value'
      ]}
      committed={true}
      path={[ '/' ]}
    />);
    const removeButtons = queryAllByTestId('remove-item-btn')
    expect(removeButtons.length).toBe(2)
    act(() => {
      fireEvent.click(removeButtons[index])
    })
    expect(removeItemMock).toHaveBeenCalledWith(index)
  })
})

export {}