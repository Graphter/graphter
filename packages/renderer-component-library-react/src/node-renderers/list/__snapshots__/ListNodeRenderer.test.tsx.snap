// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ListNodeRenderer /> should error when more than one child config is supplied 1`] = `[Error: Only one child list type is currently supported]`;

exports[`<ListNodeRenderer /> should error when no child config is supplied 1`] = `[Error: <ListNodeRenderer /> component at '/' is missing config]`;

exports[`<ListNodeRenderer /> should render correctly 1`] = `
<div>
  <div
    class="listNodeRenderer"
    data-nodepath="/"
    data-nodetype="list"
  >
    <div
      class="items"
      data-testid="items"
    >
      <div
        class="defaultItemWrapper"
        data-testid="item"
      >
        <input
          data-testid="string-renderer"
          value="item 1 value"
        />
        <button
          class="button"
          data-testid="remove-item-btn"
          type="button"
        >
          Remove [-]
        </button>
      </div>
      <div
        class="defaultItemWrapper"
        data-testid="item"
      >
        <input
          data-testid="string-renderer"
          value="item 2 value"
        />
        <button
          class="button"
          data-testid="remove-item-btn"
          type="button"
        >
          Remove [-]
        </button>
      </div>
    </div>
    <div
      class="defaultNewItemWrapper"
      data-testid="add-item"
    >
      <input
        data-testid="string-renderer"
        value=""
      />
      <button
        class="button"
        data-testid="add-item-btn"
        type="button"
      >
        Add [+]
      </button>
    </div>
  </div>
</div>
`;

exports[`<ListNodeRenderer /> should throw if non-array data is passed in 1`] = `[Error: 'list' renderer only works with arrays but got 'object']`;
